File: FreqStack.js

var FreqStack = function () {
    this.groupsWithSameFrequency = new Map();//Key: frequency | Value: array of integers with this frequency.
    this.valueToFrequency = new Map();//Key: integer | Value: frequency of this integer.
    this.maxFrequency = 0;
};

/** 
 * @param {number} value
 * @return {void}
 */
FreqStack.prototype.push = function (value) {
    let frequency = (this.valueToFrequency.has(value) ? this.valueToFrequency.get(value) : 0) + 1;
    this.valueToFrequency.set(value, frequency);
    if (!this.groupsWithSameFrequency.has(frequency)) {
        this.groupsWithSameFrequency.set(frequency, []);
    }
    this.groupsWithSameFrequency.get(frequency).push(value);
    this.maxFrequency = Math.max(this.maxFrequency, frequency);
};

/**
 * @return {number}
 */
FreqStack.prototype.pop = function () {
    let popValue = this.groupsWithSameFrequency.get(this.maxFrequency).pop();

    if (this.valueToFrequency.get(popValue) === 1) {
        this.valueToFrequency.delete(popValue);
    } else {
        this.valueToFrequency.set(popValue, this.valueToFrequency.get(popValue) - 1);
    }

    if (this.groupsWithSameFrequency.get(this.maxFrequency).length === 0) {
        this.groupsWithSameFrequency.delete(this.maxFrequency);
        this.maxFrequency--;
    }
    return popValue;
};

File: FreqStack.java

import java.util.Deque;
import java.util.ArrayDeque;
import java.util.Map;
import java.util.HashMap;

public class FreqStack {

    Map<Integer, Deque<Integer>> groupsWithSameFrequency;
    Map<Integer, Integer> valueToFrequency;
    int maxFrequency;

    public FreqStack() {
        groupsWithSameFrequency = new HashMap();
        valueToFrequency = new HashMap<>();
    }

    public void push(int value) {
        int frequency = valueToFrequency.getOrDefault(value, 0) + 1;
        valueToFrequency.put(value, frequency);
        groupsWithSameFrequency.computeIfAbsent(frequency, values -> new ArrayDeque<>()).push(value);
        maxFrequency = Math.max(maxFrequency, frequency);
    }

    public int pop() {
        int popValue = groupsWithSameFrequency.get(maxFrequency).pop();

        if (valueToFrequency.get(popValue) == 1) {
            valueToFrequency.remove(popValue);
        } else {
            valueToFrequency.put(popValue, valueToFrequency.get(popValue) - 1);
        }

        if (groupsWithSameFrequency.get(maxFrequency).isEmpty()) {
            groupsWithSameFrequency.remove(maxFrequency);
            maxFrequency--;
        }

        return popValue;
    }
}
